Requests page

klakhov opened the pull request:
<!-- Raise an issue to propose your change (https://github.com/opencv/cvat/issues).
It helps to avoid duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [Contribution guide](https://opencv.github.io/cvat/docs/contributing/). -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->

### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable for some reason, then ~~explicitly strikethrough~~ the whole
line. If you don't do that, GitHub will show incorrect progress for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [x] I have created a changelog fragment <!-- see top comment in CHANGELOG.md -->
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] I have linked related issues (see [GitHub docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary
  ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
  [cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning),
  [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and
  [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

### Release Notes

- **New Features**
  - Enhanced progress tracking for dataset export and backup operations.
  - Improved import dataset progress visibility with dynamic messages.

- **Bug Fixes**
  - Fixed navigation flow in multiple Cypress test scripts by adding `cy.goBack()` after export operations.
  - Corrected notification handling in various test cases.

- **Refactor**
  - Removed unused state properties and hooks related to `importing` and `backupIsActive` across multiple components.
  - Restructured state management in reducers for export and import actions.

- **Tests**
  - Updated Cypress commands for downloading exports and verifying downloads.
  - Added new commands for better test flow management in Cypress scripts.

- **Documentation**
  - Updated copyright year in various components.

- **Chores**
  - Added new error handling and notification messages for request-related actions.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Marishka17 commented:
/check

Marishka17 commented:
/check

Marishka17 commented:
/check

Marishka17 commented:
/check

zhiltsov-max commented on cvat/apps/engine/serializers.py:
```diff
@@ -2170,3 +2177,120 @@ def _update_assets(guide):
+    finished_date = serializers.DateTimeField(
+        required=False, allow_null=True, source="ended_at", read_only=True
+    )
+    expire_date = serializers.SerializerMethodField()
```
```suggestion
    expiry_date = serializers.SerializerMethodField()
```
Or `expiration`, both would be more correct from the language viewpoint.

zhiltsov-max commented on cvat/apps/engine/views.py:
```diff
@@ -3156,3 +3182,192 @@ def _import_project_dataset(request, rq_id_template, rq_func, db_obj, format_nam
+    list=extend_schema(
+        summary='List RQ jobs',
+        parameters=[
+            OpenApiParameter(name='task_id', location=OpenApiParameter.QUERY, type=OpenApiTypes.INT),
+            OpenApiParameter(name='action', location=OpenApiParameter.QUERY, type=OpenApiTypes.STR, enum=['create']),
```
This list of filters looks incomplete, comparing with the supported targets and actions. What do you think about adding filters like `project_id`, `job_id`, and other actions and subjects (e.g. backup, ...)? Also, what do you think about adding `status`, and time filters, sorting?

zhiltsov-max reviewed (commented):
@Marishka17, I think, it makes sense to add named constants or use other means of encapsulation for shared strings, such as `meta['user']`, actions, etc. Currently, the fact that they are the same in different jobs is coincidental.

Marishka17 commented on cvat/apps/engine/urls.py:
```diff
@@ -24,6 +24,7 @@
 router.register('cloudstorages', views.CloudStorageViewSet)
 router.register('assets', views.AssetsViewSet)
 router.register('guides', views.AnnotationGuidesViewSet)
+router.register('requests', views.RequestViewSet,  basename="request")
```
I'm not sure that I understand which pattern you mean. Are you talking about using plural instead of singular? In that case, looks like the standard pattern is `ObjectViewSet`..

zhiltsov-max replied:
Yes, I meant changing it to `RequestsViewSet`, but it seems it was my bad.

Marishka17 commented on cvat/apps/engine/views.py:
```diff
@@ -3156,3 +3182,192 @@ def _import_project_dataset(request, rq_id_template, rq_func, db_obj, format_nam
+
+    @method_decorator(never_cache)
+    @_handle_redis_exceptions
+    def destroy(self, request, pk: str):
```
I've added this endpoint just because without it user cannot "directly" delete finished/failed jobs. That means that after some time the functionality for managing requests can become useless because it will require iterating over a huge amount of jobs and finding the required one.

zhiltsov-max replied:
Well, typically, we use filters and sorting for this. Which are suggested in another comment.

klakhov commented:
/check

klakhov commented:
/check

klakhov commented:
/check

SpecLad commented on cvat/schema.yml:
```diff
@@ -9222,6 +9354,107 @@ components:
+          type: string
+          format: date-time
+          readOnly: true
+        start_date:
```
`started_date` would be more consistent.

SpecLad commented on cvat/schema.yml:
```diff
@@ -9222,6 +9354,107 @@ components:
+          format: date-time
+          readOnly: true
+          nullable: true
+        expire_date:
```
```suggestion
        expires_date:
```

SpecLad commented on cvat/schema.yml:
```diff
@@ -9222,6 +9354,107 @@ components:
+          readOnly: true
+          nullable: true
+      required:
+      - operation
```
This list seems a bit barren. What about `id`, `status`, `enqueue_date`, `owner` - shouldn't these be required too?

SpecLad commented on cvat/schema.yml:
```diff
@@ -9222,6 +9354,107 @@ components:
+      - queued
+      - started
+      - failed
+      - finished
```
If the intent here is for this to be the opposite of `failed`, then `succeeded` would be a more natural choice.

Marishka17 replied:
No, the intent here is to provide which rq statuses are supported by the server

SpecLad commented on cvat/schema.yml:
```diff
@@ -9222,6 +9354,107 @@ components:
+          type: string
+          readOnly: true
+          nullable: true
+        name:
```
What's this supposed to be?

Marishka17 replied:
This param will be removed, it was presented in the original API design, but I was waiting for the client update to stop using it.

SpecLad commented on cvat/schema.yml:
```diff
@@ -9222,6 +9354,107 @@ components:
+        type:
+          type: string
+          readOnly: true
+        target:
```
Is this field really necessary? Seems like it should be possible to get the same information from the `id`s that are present.

Marishka17 replied:
I agree with you that it is kind of sending duplicated data. The same applies to the `type` field.

SpecLad commented on cvat/schema.yml:
```diff
@@ -5403,6 +5511,9 @@ paths:
   /api/tasks/{id}/status:
     get:
       operationId: tasks_retrieve_status
+      description: 'This method is deprecated and will be removed in version 2.14.0.
```
If this is deprecated, how about updating the SDK so that it doesn't use it anymore?

Marishka17 commented on cvat/schema.yml:
```diff
@@ -9222,6 +9354,107 @@ components:
+          format: double
+          readOnly: true
+          nullable: true
+        enqueue_date:
```
I have just taken the API interface which was discussed and approved on HLD

bsekachev replied:
I will vote for `enqueued_date`.
In django admin we have `started/queued`. I am not sure this is the same time in case with RQ jobs. 
![image](https://github.com/cvat-ai/cvat/assets/40690378/f703857d-0d0e-4894-9c7b-eef103af4c97)

zhiltsov-max commented on cvat/schema.yml:
```diff
@@ -4281,6 +4281,114 @@ paths:
+      summary: List RQ jobs
+      parameters:
+      - in: query
+        name: action
+        schema:
+          type: string
+          enum:
+          - create
```
I think it's a filter

klakhov commented:
@Marishka17  
Could you check permissions for the lambda models please? It seems UI tests are failiing because of consistent error `could not get models from the server`.

bsekachev commented:
@klakhov 

- Please, merge the develop branch.
- I can't create any task. The server was restarted. `/requests/` endpoint works. Did I miss something additional? 
![image](https://github.com/cvat-ai/cvat/assets/40690378/d1ed6a54-cd61-4270-8d6f-15f98d9b705b)

bsekachev commented:
The same thing on `/tasks` page. Requests are with code 200. 

![image](https://github.com/cvat-ai/cvat/assets/40690378/4303b690-9abd-482d-a346-62f5fcb58cda)

bsekachev commented:
There is seven elements on the first page and one element on the second
Why was it implemented this way? Why not 10 per page? Empty page at the bottom allows to have at least one element more even without scrolling 

![image](https://github.com/cvat-ai/cvat/assets/40690378/ab0c4b4b-c93b-4e07-8f11-7db09e32e036)
![image](https://github.com/cvat-ai/cvat/assets/40690378/066cfa30-ec12-415c-8b84-33f6ef56c653)

bsekachev commented:
- I can't see automatic annotation request on requests page
![image](https://github.com/cvat-ai/cvat/assets/40690378/0cb3f854-dc0c-4a66-b16a-2a0032f6b1c7)

bsekachev commented:
When backup a task, first it says me that downloading will start automatically.
After that it suggest me to open another page to start downloading. Why?  

The style of these two notifications is different (font size, i mean) 

![image](https://github.com/cvat-ai/cvat/assets/40690378/374412ed-b1f5-4f6a-b1eb-34c3462cb3a7)

bsekachev commented:
Additionally, I will recommend to start downloading automatically whenever it is possible. 
Without necessity to open another page and click something else.

bsekachev commented:
Do we really store backup only 12 minutes?

![image](https://github.com/cvat-ai/cvat/assets/40690378/e4bf7d96-dff0-49ee-b09c-1041df2f5250)


All started/enqueued are equal to page refresh time:
![image](https://github.com/cvat-ai/cvat/assets/40690378/b1c8a5f7-8e8b-4586-a7fe-16e197708435)

bsekachev commented:
Why does it suggest me to DELETE if the method is not allowed? 

![image](https://github.com/cvat-ai/cvat/assets/40690378/f14b48b4-c44d-4beb-8d80-b457218aea34)

Marishka17 commented:
> Why does it suggest me to DELETE if the method is not allowed?

@bsekachev, I removed DELETE endpoint on the server, but forgot to remove corresponding action on UI. @klakhov, API has been changed on the server a little bit, as was suggested in the comments above, could you please check that and adapt UI?

Marishka17 commented:
> I can't see automatic annotation request on requests page

This was not mentioned in the HLD as will be implemented. Right now server handles only import/export queues. I can add automatic annotation queue after vacation.

bsekachev commented:
@Marishka17 

Alright. Current HLD only describes the first iteration. 
But in this case, I expect @klakhov to track the following iterations of the feature and prepare new updated HLD.

bsekachev commented:
@klakhov 

Too many requests. We can't afford this patch on production server. 
It sends requests to often (each 2 seconds...?? )

![image](https://github.com/cvat-ai/cvat/assets/40690378/1bad3242-63dc-41e0-88f3-bc73277cc83a)

Period should be more smart. For example there is not any reason to ask often the status if rq job is deferred or queued. 
If progress increases slowly, we also may increase pooling period. 
Please, think how to implement it in better way.

bsekachev commented:
It did not update status on `Requests` page.
All these tasks are already done (And I can see them to be done in a new tab, please see preview of another tab):
![image](https://github.com/cvat-ai/cvat/assets/40690378/adc6e8fa-ec77-40d4-b3b4-9d4a724f071b)

Request also has shown that the process has finished (see task id and corresponding line on `requests` page):
![image](https://github.com/cvat-ai/cvat/assets/40690378/88c38e76-8bc8-47d6-9bf0-c23cd6ab272b)

bsekachev commented:
What does expire for Import annotations job? 
![image](https://github.com/cvat-ai/cvat/assets/40690378/2ce3a82c-c936-4f9a-ab4e-bc8aa333f876)

bsekachev commented:
Why do we duplicate any exception messages?
![image](https://github.com/cvat-ai/cvat/assets/40690378/7cbc0f06-9041-4cc1-abb3-b1333d87066e)

bsekachev commented:
Long exception has to be scrollable. 
![image](https://github.com/cvat-ai/cvat/assets/40690378/81d0ff5c-1d14-4f33-8912-75e2f041a8de)

I believe the height of the card should be fixed.

bsekachev commented:
Changelog entry?

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,278 @@
+        const promise = new Promise<Request>((resolve, reject) => {
+            const timeoutCallback = (): void => {
+                serverProxy.requests.status(id, params).then((response) => {
+                    const request = new Request({ ...response });
```
Why we create `Request` instance from a variable with name `response`?

bsekachev commented:
After refreshing a page, I may request exporting at the same format again.

Before:
![image](https://github.com/cvat-ai/cvat/assets/40690378/5af6aa47-02c0-46b3-8f56-084491945fea)

After:
![image](https://github.com/cvat-ai/cvat/assets/40690378/e4c92982-8fbd-45fc-9c2f-ad74ba746819)

We need to come to consistent solution (do not disable these items at all OR disable them even if the page has been refreshed).

klakhov commented on cvat-core/src/project-implementation.ts:
```diff
@@ -96,9 +98,14 @@
         customName?: string,
+        options?: { requestStatusCallback?: (request: Request) => void },
     ) {
-        const result = exportDataset(this, format, saveImages, useDefaultSettings, targetStorage, customName);
-        return result;
+        const { requestStatusCallback } = options || {};
+        const rqID = await exportDataset(this, format, saveImages, useDefaultSettings, targetStorage, customName);
+        if (rqID) {
+            return requestsManager.listen(rqID, { callback: requestStatusCallback });
+        }
+        return new Request({ status: RQStatus.FINISHED, message: '' });
```
We dont want to use ternary operator here as the line is hard to read

klakhov commented on cvat-core/src/server-proxy.ts:
```diff
@@ -2429,6 +2246,54 @@
     }
 }
 
+async function getRequestsList(): Promise<PaginatedResource<SerializedRequest>> {
+    const { backendAPI } = config;
+
+    try {
+        const response = await fetchAll(`${backendAPI}/requests`);
+
+        return response.results;
+    } catch (errorData) {
+        throw generateError(errorData);
+    }
+}
```
We dont need additional information here

klakhov commented:
For UI part:
- Task status requests were fixed, now task creation displays normally
- Imroved amount of requests on the page, now it looks like tasks list(10 per page with scroll)
- Fixed font size on notifications
- Fixed timestamps on the requests cards + hidden the expiration time for tasks create and import
- Removed delete option
- Improved status requests delay system (now only 5 consecutive requests are allowed, others will be re-scheduled)
- Improved error display for the card, added scrollable element if its too large
- Disabled requesting export if its already started(as it was before)
- As for downloading automatically, as I remember we discussed on HLD that we should download only by button click

bsekachev commented:
I ran several processes as on the screen, than pressed F5 to refresh the page. After refresh: 
![image](https://github.com/cvat-ai/cvat/assets/40690378/5c29f30a-5a25-424f-892e-8baa366db5d9)

bsekachev commented:
![image](https://github.com/cvat-ai/cvat/assets/40690378/654b9a74-def9-4a93-99e1-fc52cad6a1f6)

1. Download once
2. Wait for a 30 seconds / one minute
3. Try to download one more time

It runs new export: 
![image](https://github.com/cvat-ai/cvat/assets/40690378/2b178a68-9fc2-47c9-ac82-51df9b26edf3)

bsekachev commented on cvat-core/src/annotations.ts:
```diff
@@ -161,20 +162,23 @@ export function importDataset(
         convMaskToPoly?: boolean,
-        updateStatusCallback?: (s: string, n: number) => void,
+        requestStatusCallback?: (request: Request) => void,
+        uploadStatusCallback?:(s: string, n: number) => void,
```
```suggestion
        uploadStatusCallback?: (s: string, n: number) => void,
```

bsekachev commented on cvat-core/src/annotations.ts:
```diff
@@ -161,20 +162,23 @@ export function importDataset(
     const convMaskToPoly = 'convMaskToPoly' in options ? options.convMaskToPoly : true;
     const adjustedOptions = {
-        updateStatusCallback,
+        requestStatusCallback,
```
We need a check for this new argument like for `uploadStatusCallback`
Message should be more specific if we have many callbacks now:

``throw new ArgumentError('Callback must be a function');`` ->  ``throw new ArgumentError('Request status callback must be a function');``

bsekachev commented on cvat-core/src/api.ts:
```diff
@@ -374,6 +375,27 @@ function build(): CVATCore {
             },
         },
+        requests: {
+            async list(): Promise<{
+                requests: Request[];
+                count: number;
+            }> {
```
For the most of resources we have returning lists like: PaginatedResource<Resource> where `count` is an additional property of returning array.

But for lambda manager and request manager we use another approach.

I believe it should be unified.

bsekachev commented on cvat-core/src/enums.ts:
```diff
@@ -50,6 +50,7 @@ export enum RQStatus {
     FINISHED = 'finished',
     FAILED = 'failed',
     UNKNOWN = 'unknown',
+    DEFERRED = 'deferred',
```
Did you test it?

I have several deferred requests, but they are returned as queued
![image](https://github.com/cvat-ai/cvat/assets/40690378/7ce84dee-a822-403c-961b-70449a5c9bad)

Marishka17 replied:
> I have several deferred requests, but they are returned as queued

This is consciously changed on the server, a user shouldn't know that his job is deferred. For him job is just in a queue.

bsekachev replied:
> a user shouldn't know that his job is deferred

1. Why do you think so? 
2. For a user on UI we will still may show "Queued", no problem.
3. But this information is required to implement correct long-time-pooling algorithm  (when user has 500+ jobs queued, will client ask the server about each status every N seconds??? Definitely No -> throttling issues. So, for which Requests should client fetch the information? I would say that we not even should now which jobs are deferred [at minimum], but also information about dependend jobs is important)

bsekachev replied:
Okay, let's postpone this algorithm, but in this case please remove all 'deferred' from cvat-core

bsekachev commented on cvat-core/src/index.ts:
```diff
@@ -150,6 +152,19 @@ export default interface CVATCore {
+    requests: {
+        list: () => Promise<{ requests: Request[], count: number }>;
+        listen: (
+            rqID: string | null,
```
What does listening for `null` request mean?

bsekachev commented on cvat-core/src/index.ts:
```diff
@@ -150,6 +152,19 @@ export default interface CVATCore {
+        listen: (
+            rqID: string | null,
+            options?: {
+                callback?: (request: Request) => void,
```
Shall this callback be optional? What is a reason to listen for something if we do not want to get updates?

klakhov replied:
Why not, if some code wants to listen but dont need to update any UI

bsekachev replied:
Could you please provide example?

bsekachev commented on cvat-core/src/index.ts:
```diff
@@ -150,6 +152,19 @@ export default interface CVATCore {
+            rqID: string | null,
+            options?: {
+                callback?: (request: Request) => void,
+                filter?: RequestsFilter,
```
What are you going to filter if you listen for a specific request?

bsekachev commented on cvat-core/src/project-implementation.ts:
```diff
@@ -96,9 +98,14 @@ export default function implementProject(projectClass) {
-        return result;
+        const { requestStatusCallback } = options || {};
+        const rqID = await exportDataset(this, format, saveImages, useDefaultSettings, targetStorage, customName);
+        if (rqID) {
+            return requestsManager.listen(rqID, { callback: requestStatusCallback });
+        }
+        return new Request({ status: RQStatus.FINISHED, message: '' });
```
I believe the following approach is more correct:

On `cvat-core` level we do not want to run implicit listening inside `exportDataset`, `importDataset`, `exportBackup`, `importBackup`. These methods should return `Request` instance. If `cvat-ui` wants to listen for them, it calls `listen` explicitly. This way we will get rid of new options, like `requestStatusCallback` and it will make code easier. 

Need to think additionally about task creating. For simplicity, my basic idea is to split between `create` and `save`  (as the first one is worker process, the second one is quicker):
- `create` returns `Request` instance.
- `save` returns `Task` instance as before.

It is not unified with for example Project.save or with Issue.save. But maybe we do not need this unification as these processes are not performed in Workers.

bsekachev replied:
For now, never mind, let's discuss after full code review, when I have mode understanding

klakhov replied:
Implemented as we discussed

bsekachev commented on cvat-core/src/project-implementation.ts:
```diff
@@ -96,9 +98,14 @@ export default function implementProject(projectClass) {
     ) {
-        const result = exportDataset(this, format, saveImages, useDefaultSettings, targetStorage, customName);
-        return result;
+        const { requestStatusCallback } = options || {};
```
`requestStatusCallback` type is not checked

bsekachev commented on cvat-core/src/project-implementation.ts:
```diff
@@ -107,24 +114,37 @@ export default function implementProject(projectClass) {
         },
     ) {
-        return importDataset(this, format, useDefaultSettings, sourceStorage, file, options);
+        const { requestStatusCallback } = options || {};
```
`requestStatusCallback` type is not checked

bsekachev commented on cvat-core/src/project-implementation.ts:
```diff
@@ -107,24 +114,37 @@ export default function implementProject(projectClass) {
     ) {
-        const result = await serverProxy.projects.backup(this.id, targetStorage, useDefaultSettings, fileName);
-        return result;
+        const { requestStatusCallback } = options || {};
```
`requestStatusCallback` type is not checked

bsekachev commented on cvat-core/src/project-implementation.ts:
```diff
@@ -107,24 +114,37 @@ export default function implementProject(projectClass) {
-    projectClass.restore.implementation = async function (storage: Storage, file: File | string) {
-        const result = await serverProxy.projects.restore(storage, file);
-        return result;
+    projectClass.restore.implementation = async function (
```
`requestStatusCallback` type is not checked

bsekachev commented on cvat-core/src/project-implementation.ts:
```diff
@@ -96,9 +98,14 @@ export default function implementProject(projectClass) {
-        const result = exportDataset(this, format, saveImages, useDefaultSettings, targetStorage, customName);
-        return result;
+        const { requestStatusCallback } = options || {};
+        const rqID = await exportDataset(this, format, saveImages, useDefaultSettings, targetStorage, customName);
+        if (rqID) {
+            return requestsManager.listen(rqID, { callback: requestStatusCallback });
+        }
```
Basically `exportDataset`, from this implementaiton will always return rq id or throw an exception.
So, we do not need this `if` condition. Just update typings for `exportDataset`

bsekachev commented on cvat-core/src/project-implementation.ts:
```diff
@@ -107,24 +114,37 @@ export default function implementProject(projectClass) {
-        const result = await serverProxy.projects.backup(this.id, targetStorage, useDefaultSettings, fileName);
-        return result;
+        const { requestStatusCallback } = options || {};
+        const rqID = await serverProxy.projects.backup(this.id, targetStorage, useDefaultSettings, fileName);
+        if (rqID) {
+            return requestsManager.listen(rqID, { callback: requestStatusCallback });
+        }
```
The same comment as for `exportDataset`

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,357 @@
+        return this.#owner;
+    }
+
+    updateFields(request: Partial<SerializedRequest>): void {
+        if (request.id !== undefined) {
+            this.#id = request.id;
+        }
+        if (request.status !== undefined) {
+            this.#status = request.status as RQStatus;
+        }
+        if (request.operation !== undefined) {
+            this.#operation = request.operation;
+        }
+        if (request.progress !== undefined) {
+            this.#progress = request.progress;
+        }
+        if (request.message !== undefined) {
+            this.#message = request.message;
+        }
+        if (request.result_url !== undefined) {
+            this.#resultUrl = request.result_url;
+        }
+        if (request.enqueued_date !== undefined) {
+            this.#enqueuedDate = request.enqueued_date;
+        }
+        if (request.started_date !== undefined) {
+            this.#startedDate = request.started_date;
+        }
+        if (request.finished_date !== undefined) {
+            this.#finishedDate = request.finished_date;
+        }
+        if (request.expiry_date !== undefined) {
+            this.#expiryDate = request.expiry_date;
+        }
+        if (request.owner !== undefined) {
+            this.#owner = new User(request.owner);
+        }
+    }
+}
```
Why do not implement setters? The method makes sense to reduce code duplication, but in this case another question:

- We have setters for `status` and `message`, but not for others. Why?

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,357 @@
+    async list(): Promise<{ requests: Request[], count: number }> {
+        const result = await serverProxy.requests.list();
+        const requests = [];
+        for (const request of result) {
+            requests.push(
+                new Request({
+                    ...request,
+                }),
+            );
+        }
```
`map`?

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,357 @@
+            initialRequest?: Request,
+        },
+    ): Promise<Request> {
+        let storedID = null;
```
Why `storedID`? I will suggest to rename to `requestID`

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,357 @@
+        const promise = new Promise<Request>((resolve, reject) => {
+            const timeoutCallback = (): void => {
+                if (!(storedID in this.listening)) {
+                    return;
```
How is that possible if you clear timeout before deleting the request from this.listening?

klakhov replied:
Removed

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,357 @@
+                }
+                this.requestStack.unshift(timestamp);
+
+                serverProxy.requests.status(id, params).then((serializedRequest) => {
```
I will suggest to rewrite with `async/await`.

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,357 @@
+                        reject(error);
+                    }
+                }).finally(() => {
+                    if (storedID in this.listening) {
+                        this.listening[storedID].timeout = null;
+                    }
```
In then you have:

```
this.listening[storedID].timeout = window.setTimeout(timeoutCallback, this.delayFor(storedID));
```

In finally you set it to `null`. Why?

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,357 @@
+                    if (storedID in this.listening) {
+                        const { onUpdate } = this.listening[storedID];
+                        if ([RQStatus.QUEUED, RQStatus.STARTED, RQStatus.DEFERRED].includes(status)) {
+                            onUpdate.forEach((update) => update(request));
```
We can't guarantee one of update callback do not change the request object.

In this case, I will suggest different ways to ensure that it does not happen:
1. Do Request object fully read only
2. Create new Request instance each time for each callback
3. Using typescript define that this argument passes as read only

klakhov replied:
Made the Request read-only

bsekachev commented on cvat-ui/src/actions/requests-async-actions.ts:
```diff
@@ -0,0 +1,165 @@
+            initialRequest: request,
+        })
+        .catch((error: Error) => {
+            request.updateFields({ status: RQStatus.UNKNOWN, progress: 0, message: '' });
```
What is the reason to  update fields outside of cvat-core?

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -855,24 +856,16 @@ function exportDataset(instanceType: 'projects' | 'jobs' | 'tasks') {
-                            resolve(`${baseURL}?${new URLSearchParams(params).toString()}`);
-                        } else if (isCloudStorage && status === 200) {
-                            resolve();
+                        if (response.status === 202) {
+                            resolve(response.data.rq_id);
                         }
+                        resolve();
```
Are there any cases, when success response does not include `rq_id`?

klakhov replied:
Yes, in case of export and backups, if the export is finished but not downloaded, new export is not started and we get 200 response without `rq_id` (the same thing for other same comments)

bsekachev replied:
So, then something is not working here.
For example:

- serverProxy.backupTask or backupProject returns string or undefined.
- it means Task.backup, and Project.backup also returns string or undefined
- string or undefined is used to start listening (in case with undefined, it seems there is typing error)

![image](https://github.com/cvat-ai/cvat/assets/40690378/943cfa6c-a1d0-427b-a61d-496f886ee93e)

bsekachev replied:
Perhaps the same for export, but I did not check

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -972,13 +941,18 @@ async function importDataset(
     }
     try {
-        return await wait();
+        return rqId;
```
I do not think we need try catch here :)

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -907,30 +900,6 @@ async function importDataset(
-    }
     const isCloudStorage = sourceStorage.location === StorageLocation.CLOUD_STORAGE;
 
     if (isCloudStorage) {
```
We do not need `if .. else`` here anymore

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -987,23 +961,16 @@ async function backupTask(id: number, targetStorage: Storage, useDefaultSettings
-                    resolve(`${url}?${new URLSearchParams(params).toString()}`);
-                } else if (isCloudStorage && status === 200) {
-                    resolve();
+                if (response.status === 202) {
+                    resolve(response.data.rq_id);
                 }
+                resolve();
```
The same question about rq_id here

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -1074,15 +1018,16 @@ async function restoreTask(storage: Storage, file: File | string) {
             });
     }
-    return wait();
+    const rqId = response.data.rq_id;
```
Here we also do not really need ``if ... else`` anymore

Just: `if (isCloudStorage) { response = ..; return response.data.rq_id }` 

By the way I noticed that requests are not try..catched here.

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -1093,23 +1038,16 @@ async function backupProject(
-                } else if (isCloudStorage && status === 200) {
-                    resolve();
+                if (response.status === 202) {
+                    resolve(response.data.rq_id);
```
The same question about rq_id

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -1182,92 +1094,21 @@ async function restoreProject(storage: Storage, file: File | string) {
     }
-    return wait();
+
+    const rqId = response.data.rq_id;
```
The same comment about if ... else as before

bsekachev replied:
And about exception handling

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -1380,8 +1221,7 @@ async function createTask(
     try {
-        const createdTask = await listenToCreateTask(response.data.id, onUpdate);
-        return createdTask;
+        return response.data.id;
```
Please, remove try .. catch.

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -1817,7 +1634,7 @@ async function uploadAnnotations(
     }
     try {
-        return await wait();
+        return rqId;
```
1. Please, remove try catch from return value
2. Please, refactor if/else for cloud storage / not a cloud storage

bsekachev commented on cvat-core/src/session-implementation.ts:
```diff
@@ -506,10 +525,11 @@ export function implementTask(Task) {
     ): Promise<TaskClass> {
         if (Number.isInteger(this.id) && this.size === 0) {
-            const serializedTask = await serverProxy.tasks.listenToCreate(this.id, onUpdate);
+            const request = await requestsManager.listen(null, options);
```
Listen for null ????????

bsekachev commented on cvat-ui/src/components/tasks-page/task-item.tsx:
```diff
@@ -58,17 +59,21 @@ class TaskItemComponent extends React.PureComponent<TaskItemProps & RouteCompone
-                    });
-                }
-            }).then((createdTask: Task) => {
+            taskInstance.listenToCreate({
+                callback: (request: Request) => {
+                    if (!this.#isUnmounted) {
+                        this.setState({
+                            importingState: {
+                                message: request.message,
+                                progress: Math.floor(request.progress * 100),
+                                state: request.status,
+                            },
+                        });
+                    }
+                },
+                filter: { taskID: taskInstance.id, action: 'create' },
+            },
+            ).then((createdTask: Task) => {
```
Why do not we have normal `rq_id` for a task creation process? 
This implementation with additional option filter and null rq_id does not look pretty

bsekachev commented on cvat-core/src/session-implementation.ts:
```diff
@@ -331,9 +333,14 @@ export function implementJob(Job) {
-        return result;
+        const { requestStatusCallback } = options || {};
+        const rqID = await exportDataset(this, format, saveImages, useDefaultSettings, targetStorage, customName);
+        if (rqID) {
+            return requestsManager.listen(rqID, { callback: requestStatusCallback });
+        }
+        return new Request({ status: RQStatus.FINISHED, message: '' });
```
Export dataset should always return normal rqID or throw Error

bsekachev commented on cvat-core/src/session-implementation.ts:
```diff
@@ -530,15 +550,24 @@ export function implementTask(Task) {
-        return result;
+        const { requestStatusCallback } = options || {};
+        const rqID = await serverProxy.tasks.backup(this.id, targetStorage, useDefaultSettings, fileName);
+        if (rqID) {
+            return requestsManager.listen(rqID, { callback: requestStatusCallback });
+        }
+        return new Request({ status: RQStatus.FINISHED, message: '' });
```
Backup should always return normal rqID or throw Error

bsekachev commented on cvat-core/src/session-implementation.ts:
```diff
@@ -783,9 +813,14 @@ export function implementTask(Task) {
-        return result;
+        const { requestStatusCallback } = options || {};
+        const rqID = await exportDataset(this, format, saveImages, useDefaultSettings, targetStorage, customName);
+        if (rqID) {
+            return requestsManager.listen(rqID, { callback: requestStatusCallback });
+        }
+        return new Request({ status: RQStatus.FINISHED, message: '' });
```
Export dataset should always return normal rqID or throw Error

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,313 @@
+    get message(): string {
+        return this.#message;
+    }
+    set message(message) {
+        this.#message = message;
+    }
```
How do you use it?

klakhov replied:
Removed

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,313 @@
+    get status(): RQStatus {
+        return this.#status.toLowerCase() as RQStatus;
+    }
+    set status(status) {
+        this.#status = status;
+    }
```
How do you use it?

bsekachev commented on cvat-core/src/server-proxy.ts:
```diff
@@ -1128,42 +1068,22 @@ async function restoreProject(storage: Storage, file: File | string) {
+                    params,
+                });
+            return response.data.rq_id;
+        } catch (errorData) {
+            throw generateError(errorData);
+        }
```
Since `catch` is the same, I will recommend to use common try in the function body. Please, check other changed functions in this file likewise

bsekachev commented on cvat-core/src/session-implementation.ts:
```diff
@@ -491,7 +492,15 @@ export function implementTask(Task) {
+            taskDataSpec,
+            options?.uploadStatusCallback || (() => {}),
+        );
+        await requestsManager.listen(rqID, {
+            callback: options?.requestStatusCallback,
+        });
```
Did you decide to generate server id on client and use common endpoint?

klakhov replied:
Yes, I think it has advantages: We have guessing logic in one place - serverProxy, so we dont need to invent client ids and then parse it to send filter request.  
If something changes on server, we can change one line on the client(but I suppose next change on the server should be returning correct id)

bsekachev replied:
Okay

bsekachev commented on cvat-ui/src/actions/requests-actions.ts:
```diff
@@ -0,0 +1,106 @@
+    GET_REQUESTS = 'GET_REQUESTS',
+    GET_REQUESTS_SUCCESS = 'GET_REQUESTS_SUCCESS',
+    GET_REQUESTS_FAILED = 'GET_REQUESTS_FAILED',
+    GET_REQUESTS_STATUS_SUCCESS = 'GET_REQUESTS_STATUS_SUCCESS',
+    GET_REQUESTS_STATUS_FAILED = 'GET_REQUESTS_STATUS_FAILED',
```
```suggestion
    GET_REQUEST_STATUS_SUCCESS = 'GET_REQUEST_STATUS_SUCCESS',
    GET_REQUEST_STATUS_FAILED = 'GET_REQUEST_STATUS_FAILED',
```

bsekachev commented on cvat-ui/src/actions/requests-actions.ts:
```diff
@@ -0,0 +1,106 @@
+
+import { ActionUnion, createAction } from 'utils/redux';
+import { RequestsQuery } from 'reducers';
+import {
+    Request, InstanceType, getCore,
+} from 'cvat-core-wrapper';
```
```suggestion
import { Request, InstanceType, getCore } from 'cvat-core-wrapper';
```

bsekachev commented:
Minor bug:
1. Try to backup project, wait until done
![image](https://github.com/cvat-ai/cvat/assets/40690378/d12eeac5-a9fe-4a05-bed7-259a83f8ab4a)

2. Try to backup it one more time. You immediately will get two messages. One of them says it was exported to the cloud storage
![image](https://github.com/cvat-ai/cvat/assets/40690378/97f417cc-d005-4e99-a533-3d37bd7fc960)

But export was local

bsekachev commented:
It looks like the feature to delete requests absent in the implementation

![image](https://github.com/cvat-ai/cvat/assets/40690378/983497b6-4b3c-4b0f-a83f-cecfd0e670f3)

the method `core.requests.delete` exists in the cvat-core, but I can't see it somewhere to be used.

bsekachev commented on cvat-ui/src/components/requests-page/styles.scss:
```diff
@@ -0,0 +1,175 @@
+}
+
+.cvat-requests-card {
+    height: $grid-unit-size * 11;
```
96px looks better

```suggestion
    height: $grid-unit-size * 12;
```

bsekachev replied:
![image](https://github.com/cvat-ai/cvat/assets/40690378/2011089e-48ec-4e87-a0e2-bacac4ec5664)
![image](https://github.com/cvat-ai/cvat/assets/40690378/0cc10342-4dae-4fa2-9d5e-9c55ae2ee81f)

bsekachev commented on cvat-ui/src/reducers/import-reducer.ts:
```diff
@@ -1,22 +1,22 @@
-export function defineActititiesField(instance: any): 'projects' | 'tasks' | 'jobs' {
-    if (instance instanceof core.classes.Project) {
+export function defineActititiesField(instance: InstanceType | RequestInstanceType): 'projects' | 'tasks' | 'jobs' {
+    const instanceType = getInstanceType(instance);
```
```
return `${getInstanceType(instance)}s`; 
```

bsekachev commented on cvat-ui/src/actions/tasks-actions.ts:
```diff
@@ -258,16 +259,38 @@ ThunkAction<Promise<void>, {}, {}, AnyAction> {
-                } else if ([RQStatus.QUEUED, RQStatus.STARTED].includes(status)) {
-                    const helperMessage = 'You may close the window.';
+            const savedTask = await taskInstance.save({
+                uploadStatusCallback: (status: RQStatus, progress: number, message: string): void => {
+                    if (status === RQStatus.UNKNOWN) {
+                        onProgress?.(`${message} ${progress ? `${Math.floor(progress * 100)}%` : ''}`);
+                    } else if ([RQStatus.QUEUED, RQStatus.STARTED].includes(status)) {
+                        const helperMessage = 'You may close the window.';
+                        onProgress?.(`${message} ${progress ? `${Math.floor(progress * 100)}%` : ''}. ${helperMessage}`);
+                    } else {
+                        onProgress?.(`${status}: ${message}`);
+                    }
+                },
+                requestStatusCallback(request) {
+                    let { message } = request;
+                    let helperMessage = '';
+                    const { status, progress } = request;
+                    if (!message) {
+                        if ([RQStatus.QUEUED, RQStatus.STARTED].includes(status)) {
+                            message = 'CVAT queued the task to import';
+                            helperMessage = 'You may close the window.';
+                        } else if (status === RQStatus.FAILED) {
+                            message = 'Images processing failed';
+                        } else if (status === RQStatus.FINISHED) {
+                            message = 'Task creation finished';
+                        } else {
+                            message = 'Unknown status received';
+                        }
+                    }
```
What is the difference between these two callbacks? 
Do we really need both? They look very similar

klakhov replied:
Removed

bsekachev commented on cvat-ui/src/actions/requests-async-actions.ts:
```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: MIT
+
+import { ThunkAction } from 'utils/redux';
+import {
+    CombinedState,
+    RequestsQuery, StorageLocation,
+} from 'reducers';
```
```suggestion
import { CombinedState, RequestsQuery, StorageLocation } from 'reducers';
```

bsekachev commented on cvat-ui/src/actions/requests-async-actions.ts:
```diff
@@ -0,0 +1,138 @@
+            dispatch(requestsActions.getRequestsSuccess(requests));
+
+            if (notify) {
+                const shownNotifications = JSON.parse(localStorage.getItem('requestsNotifications') || '[]');
```
0. Parsing item from local storage should be try catched, otherwise now it will get into global try catch and show "get requests failed message". That is not true. 
1. Local storage is limited by 5 MB, we will quickly make it full with big number of request (as I see you only append something inside. But never remove. So, gradually it will be full)
2. Finally, I believe we may totally omit this logic. Thus, a user will only see notification if the process finished when the page is opened. I believe it is enough.

klakhov replied:
Removed this logic

bsekachev commented on cvat-ui/src/actions/requests-async-actions.ts:
```diff
@@ -0,0 +1,138 @@
+                            if (!(state.export.tasks.backup.current?.[(instance as RequestInstanceType).id] ||
+                                state.export.projects.backup.current?.[(instance as RequestInstanceType).id])
+                            ) {
+                                dispatch(exportBackupAsync(
+                                    instance as RequestInstanceType,
+                                    undefined,
+                                    undefined,
+                                    undefined,
+                                    listeningPromise,
+                                ));
+                            }
+                        } else if (!(state.export.tasks.dataset.current?.[(instance as RequestInstanceType).id] ||
+                            state.export.projects.dataset.current?.[(instance as RequestInstanceType).id] ||
+                            state.export.jobs.dataset.current?.[(instance as RequestInstanceType).id])
+                        ) {
+                            dispatch(exportDatasetAsync(
+                                instance as RequestInstanceType,
+                                format,
+                                operationTarget === 'dataset',
+                                true,
+                                undefined,
+                                undefined,
+                                listeningPromise,
+                            ));
+                        }
+                    } else if (operationType === 'import') {
+                        if (!(state.import.tasks.dataset.current?.[(instance as RequestInstanceType).id] ||
+                            state.import.projects.dataset.current?.[(instance as RequestInstanceType).id] ||
+                            state.import.jobs.dataset.current?.[(instance as RequestInstanceType).id])
+                        ) {
+                            dispatch(importDatasetAsync(
+                                instance as RequestInstanceType,
+                                format,
+                                undefined,
+                                undefined,
+                                undefined,
+                                undefined,
+                                listeningPromise,
+                            ));
```
May you please explain logic behind this? 

It makes signature and implemenetion of `exportBackupAsync`, `exportDatasetAsync`, `importDatasetAsync` more difficult, but I can't see the reason.

klakhov replied:
Reworked as discussed, using separate listening functions

bsekachev commented on cvat-ui/src/actions/requests-actions.ts:
```diff
@@ -0,0 +1,99 @@
+            request,
+        })
+    ),
+    getRequestStatusFailed: (request: Request, error: any) => (
```
This action is never called

bsekachev commented on cvat-ui/src/actions/export-actions.ts:
```diff
@@ -61,68 +71,105 @@ export const exportActions = {
+    try {
+        let result;
+        let target: 'cloudstorage' | 'local' | undefined;
+        if (rqID) {
```
Declared type is string, but we have this condition here

klakhov replied:
Added `undefined`

bsekachev commented on cvat-ui/src/actions/requests-async-actions.ts:
```diff
@@ -0,0 +1,106 @@
+                                );
+                            }
+                        }
+                    } else if (operationType === 'import') {
```
import backup?

klakhov replied:
Added

bsekachev commented on cvat-ui/src/actions/requests-actions.ts:
```diff
@@ -0,0 +1,99 @@
+    getRequests: (query?: RequestsQuery) => createAction(RequestsActionsTypes.GET_REQUESTS, { query }),
+    requestFinished: (request: Request) => createAction(RequestsActionsTypes.REQUEST_FINISHED, { request }),
+    requestFailed: (request: Request) => createAction(RequestsActionsTypes.REQUEST_FAILED, { request }),
+    getRequestsSuccess: (requests: Request[]) => createAction(
```
Request[] is a wrong type, as you read property `count`, not existing on type Array

Correct one is `Awaited<ReturnType<typeof core['requests']['list']>>`

bsekachev commented on cvat-ui/src/actions/requests-async-actions.ts:
```diff
@@ -0,0 +1,106 @@
+
+                    if (operationType === 'export') {
+                        if (operationTarget === 'backup') {
+                            if (!(state.export.tasks.backup.current?.[(instance as RequestInstanceType).id] ||
+                                state.export.projects.backup.current?.[(instance as RequestInstanceType).id])
```
Please, check the condition.
It seems like there will be a bug if we backup both project and task with the same id.

Probably better to rewrite something like:
```
                            if (
                                (instance.type === 'task' && !(instance.id in state.export.tasks.backup.current)) ||
                                (instance.type === 'project' && !(instance.id in state.export.projects.backup.current))
                            ) {
                                // dispatch
                                // listen
                            }
```

bsekachev replied:
That is relevant for  import/export annotations/dataset
And import/export backups

bsekachev commented on cvat-ui/src/actions/requests-async-actions.ts:
```diff
@@ -0,0 +1,106 @@
+                            if (!(state.export.tasks.backup.current?.[(instance as RequestInstanceType).id] ||
+                                state.export.projects.backup.current?.[(instance as RequestInstanceType).id])
+                            ) {
+                                dispatch(exportActions.exportBackup(instance as RequestInstanceType));
```
We already discussed it, so, what is the reason to call here:

- exportActions.exportDataset
- importActions.importDataset
- exportActions.exportBackup

And generally store import/export as activities. 
Only to lock some controls on UI?

klakhov commented on cvat-ui/src/actions/import-actions.ts:
```diff
@@ -1,16 +1,19 @@
 import { CombinedState } from 'reducers';
 import {
-    getCore, Storage, Job, Task, Project,
+    getCore, Storage, Job, Task, Project, InstanceType,
 } from 'cvat-core-wrapper';
 import { EventScope } from 'cvat-logger';
 import { getProjectsAsync } from './projects-actions';
 import { AnnotationActionTypes, fetchAnnotationsAsync } from './annotation-actions';
+import {
+    getInstanceType, listen, RequestInstanceType, RequestsActions,
+} from './requests-actions';
```
We dont want to use type imports, please dont suggest that anymore

klakhov commented on cvat-ui/src/actions/requests-async-actions.ts:
```diff
@@ -0,0 +1,115 @@
+                                (target === 'project' && !(state.export.projects.backup.current?.[instanceID]))
+                            ) {
+                                dispatch(exportActions.exportBackup(instance as RequestInstanceType));
+                                listenExportBackupAsync(rqID, dispatch, { instance: instance as RequestInstanceType });
+                            }
+                        } else if (operationTarget === 'dataset' || operationTarget === 'annotations') {
+                            const field = operationTarget === 'dataset' ? 'dataset' : 'annotations';
+                            if (
+                                (target === 'task' && !(state.export.tasks[field].current?.[instanceID])) ||
+                                (target === 'project' && !(state.export.projects[field].current?.[instanceID])) ||
+                                (target === 'job' && !(state.export.jobs[field].current?.[instanceID]))
+                            ) {
+                                dispatch(exportActions.exportDataset(instance as RequestInstanceType, format, field));
+                                listenExportDatasetAsync(
+                                    rqID,
+                                    dispatch,
+                                    { instance: instance as RequestInstanceType, format, saveImages: type.includes('dataset') },
+                                );
+                            }
+                        }
+                    } else if (operationType === 'import') {
+                        if (operationTarget === 'backup') {
+                            dispatch(importActions.importBackup());
+                            listenImportBackupAsync(rqID, dispatch, { instanceType: (instance as RequestInstanceType).type as 'project' | 'task' });
+                        } else if (operationTarget === 'dataset' || operationTarget === 'annotations') {
+                            if (
+                                (target === 'task' && !(state.import.tasks.dataset.current?.[instanceID])) ||
+                                (target === 'project' && !(state.import.projects.dataset.current?.[instanceID])) ||
+                                (target === 'job' && !(state.import.jobs.dataset.current?.[instanceID]))
+                            ) {
+                                dispatch(importActions.importDataset(instance as RequestInstanceType, format));
+                                listenImportDatasetAsync(
+                                    rqID,
+                                    dispatch,
+                                    { instance: instance as RequestInstanceType },
+                                );
+                            }
+                        }
+                    }
+                });
```
We dont want to use for of here

klakhov commented on cvat-ui/src/actions/requests-actions.ts:
```diff
@@ -0,0 +1,92 @@
+    getRequestsSuccess: (requests: Request[]) => createAction(
+        RequestsActionsTypes.GET_REQUESTS_SUCCESS, { requests },
+    ),
+    getRequestsFailed: (error: any) => createAction(RequestsActionsTypes.GET_REQUESTS_FAILED, {
```
We dont want to specify Error type for now, as we need to change it all across the application.

klakhov commented:
@coderabbitai pause

bsekachev commented:
Something does not work when I am within organization. I see two problems:

1. In organization I can't see request for creating a task within the same organization
2. But I see request for creating a task that was initiated outside the organization 

@Marishka17  they are probably server-side issues.

bsekachev commented on cvat-ui/src/reducers/import-reducer.ts:
```diff
@@ -36,7 +33,12 @@ const defaultState: ImportState = {
         dataset: {
             modalInstance: null,
-            current: {},
+            uploadState: {
+                id: null,
+                format: '',
+                progress: 0,
+                status: '',
+            },
```
If it was supposed to add upload progress bar (like we have when importing a dataset to a project) to tasks and jobs
It does not work. If not, please, explain

klakhov replied:
Removed. That field is used only for projects

bsekachev commented on cvat-core/src/api.ts:
```diff
@@ -374,6 +375,26 @@ function build(): CVATCore {
+            async listen(
+                rqID: string,
+                options?: {
+                    callback?: (request: Request) => void,
```
Still not clear why callback is optional

klakhov replied:
Made it required

bsekachev commented on cvat-ui/src/components/requests-page/requests-page.tsx:
```diff
@@ -0,0 +1,58 @@
+export default function RequestsPageComponent(): JSX.Element {
+    const history = useHistory();
+    const dispatch = useDispatch();
+    const fetching = useSelector((state: CombinedState) => state.requests.fetching);
+    const count = useSelector((state: CombinedState) => state.requests.count);
+    const query = useSelector((state: CombinedState) => state.requests.query);
```
Please, minimize useSelectors

bsekachev commented on cvat-ui/src/components/requests-page/requests-page.tsx:
```diff
@@ -0,0 +1,58 @@
+    const count = useSelector((state: CombinedState) => state.requests.count);
+    const query = useSelector((state: CombinedState) => state.requests.query);
+
+    const updatedQuery = { ...query };
+    const queryParams = new URLSearchParams(history.location.search);
+    for (const key of Object.keys(updatedQuery)) {
+        (updatedQuery as Indexable)[key] = queryParams.get(key) || null;
+        if (key === 'page') {
+            updatedQuery.page = updatedQuery.page ? +updatedQuery.page : 1;
+        }
+    }
```
We only have one query parameter, the code may be significantly simpler

bsekachev commented on cvat-ui/src/components/requests-page/styles.scss:
```diff
@@ -0,0 +1,175 @@
+//
+// SPDX-License-Identifier: MIT
+
+@import '../../base';
```
```suggestion
@import 'base';
```

bsekachev commented on cvat-ui/src/components/requests-page/styles.scss:
```diff
@@ -0,0 +1,175 @@
+
+@import '../../base';
+
+.cvat-requests-scrollbar {
+    &::-webkit-scrollbar {
+        background-color: #fff;
+        width: 16px;
+    }
+
+    &::-webkit-scrollbar-track {
+        background-color: #fff;
+    }
+
+    &::-webkit-scrollbar-thumb {
+        background-color: #babac0;
+        border-radius: 16px;
+        border: 6px solid #fff;
+    }
+}
+
```
Why did you decide to use vendor-specific css selectors?

klakhov replied:
I believe we cant make pretty scrollbar with standard selectors. It works fine in chrome as its our main supported browser

bsekachev commented on cvat-ui/src/components/requests-page/styles.scss:
```diff
@@ -0,0 +1,175 @@
+.cvat-requests-page-actions-button {
+    border: none;
+    padding: 0;
+    padding-top: 22px;
```
Please, consider using $grid-unit-size in css rules

klakhov replied:
Added

bsekachev commented on cvat-ui/src/components/requests-page/styles.scss:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    .ant-collapse-header {
+        padding: 0 !important;
```
May we get rid of `important`?

klakhov replied:
Removed

bsekachev commented on cvat-ui/src/components/requests-page/requests-list.tsx:
```diff
@@ -0,0 +1,74 @@
+    const count = useSelector((state: CombinedState) => state.requests.count);
+    const query = useSelector((state: CombinedState) => state.requests.query);
+
+    const queryParams = new URLSearchParams(history.location.search);
+    const updatedQuery = { ...query };
+    for (const key of Object.keys(updatedQuery)) {
+        (updatedQuery as Indexable)[key] = queryParams.get(key) || null;
+        if (key === 'page') {
+            updatedQuery.page = updatedQuery.page ? +updatedQuery.page : 1;
+        }
+    }
```
We already have this code in `requests-page` component

bsekachev commented on cvat-ui/src/components/requests-page/requests-list.tsx:
```diff
@@ -0,0 +1,74 @@
+function RequestsList(): JSX.Element {
+    const dispatch = useDispatch();
+    const history = useHistory();
+    const requests = useSelector((state: CombinedState) => state.requests.requests);
+    const count = useSelector((state: CombinedState) => state.requests.count);
+    const query = useSelector((state: CombinedState) => state.requests.query);
```
Please, minimize useSelectors

bsekachev commented on cvat-ui/src/components/requests-page/requests-list.tsx:
```diff
@@ -0,0 +1,74 @@
+                <Pagination
+                    className='cvat-tasks-pagination'
+                    onChange={(newPage: number) => {
+                        dispatch(getRequestsAsync({
+                            ...query,
+                            page: newPage,
+                        }));
```
serverProxy:getRequestsAsync uses `fetchAll`. By the way, why???

If it already fetched all list, why do you call getRequestsAsync?

klakhov replied:
Removed that call.
Dont we need to check for all requests status? There can be more than 10 in progress. We want notifications for those when they are completed

bsekachev commented on cvat-ui/src/components/requests-page/requests-list.tsx:
```diff
@@ -0,0 +1,74 @@
+    );
+}
+
+export default RequestsList;
```
Please, add React.memo where necessary. I see it is missed in other files also. 
Otherwise these components will perform extra re-rendering.

bsekachev commented on cvat-core/src/config.ts:
```diff
@@ -17,6 +19,16 @@ const config = {
     onOrganizationChange: null,
     globalObjectsCounter: 0,
+
+    requestsStatusDelays: (window as any).Cypress ? {
```
Please, remove Cypress mention from **cvat-core** code.

bsekachev commented on cvat-core/src/requests-manager.ts:
```diff
@@ -0,0 +1,199 @@
+        const state = this.listening[rqID];
+        const { request, requestDelayIdx } = state;
+
+        // request is not checked yet, call it immideately
```
```suggestion
        // request was not checked yet, call it immediately
```

bsekachev commented on cvat-core/src/server-response-types.ts:
```diff
@@ -495,3 +495,26 @@ export interface SerializedAPISchema {
 }
+
+export interface SerializedRequest {
+    id?: string;
```
why having `id` is not guaranteed? 

tha same about `created_date` and `owner`

klakhov replied:
Yes. When we create task, we use the same request callback, so we create fake Request instances. We dont have an ID while uploading data, also it wouldnt be pretty to construct a created_date and owner here.

bsekachev commented:
- The card shows a link to a task that has been deleted 
![image](https://github.com/cvat-ai/cvat/assets/40690378/a27857e1-b312-46d5-aa91-df15db4efcbc)

- One more found problem is that in `develop` if you try to create a task and creating thread will be failed (for example try to upload a text file as data, or broken image file), UI will send a request to remove the task after failing. However with this patch - it will not, so, we may see failed tasks in the list.

bsekachev commented:
I already mentioned this, but did not get answer.
![image](https://github.com/cvat-ai/cvat/assets/40690378/bbefc592-b879-4663-8323-98e5e69e8493)

Why is TTL 20 minutes for exports??? 
Now downloading does not start automatically anymore, so, it will not work. 
If a user does not catch manually a finished job in 20 minutes, will it be necessary to export from scratch?

bsekachev commented on cvat-ui/src/components/requests-page/styles.scss:
```diff
@@ -0,0 +1,175 @@
+        height: 100%;
+    }
+
+    .ant-collapse-header {
```
As far as I see you are not  using Collapse element within request card

klakhov replied:
Removed.

klakhov commented:
- Fixed links to deleted tasks
- Added back deleting of tasks failed to create
Concerning TTL for exports, its not actually 20mins, its just a visual bug. It seems we get timestamp for expire from server in wrong format so its displayed incorrecty. I beleive Maria is already working on this.

SpecLad commented on cvat/schema.yml:
```diff
@@ -5376,6 +5825,70 @@ paths:
+      - basicAuth: []
+      responses:
+        '202':
+          description: Exporting has been started
```
This returns a request ID, right? Can you add a response schema?

SpecLad commented on cvat/schema.yml:
```diff
@@ -9274,6 +9811,115 @@ components:
+          minimum: 0
+          readOnly: true
+          nullable: true
+        created_date:
+          type: string
+          format: date-time
+          readOnly: true
+        enqueued_date:
+          type: string
+          format: date-time
+          readOnly: true
+          nullable: true
```
Is there a meaningful difference between `created_date` and `enqueued_date`, as far as the user is concerned? If a request was created, it must have also been enqueued.

Marishka17 replied:
Probably not for now, but I think it depends on whether we're going to inform the user about deferred jobs or continue to mask them as queued.

SpecLad replied:
Then let's only expose one of these fields until such a time when there's a reason to expose the other one.

(FWIW, my vote is for `created_date`, as that would be consistent with other API resources.)

klakhov commented:
I will close this PR and re-open it. It became messy with lots of comments from AI tools and reviews.
